File: IntCode/Interpreter.fs
module Interpreter

open System

type BinaryOp =
    { Lhs: int
      Rhs: int
      Destination: int }

type OpCode =
    | Add = 1
    | Multiply = 2
    | Input = 3
    | Output = 4
    | Halt = 99

type ParamMode =
    | Position = 0
    | Immediate = 1

type InterpreterAction =
    | Increment of int
    | Halt of int
    | Abort of string


let rec readUserInput (msg: string) =
    Console.Write(msg)
    let input = Console.ReadLine() |> Int32.TryParse
    match input with
    | true, num -> num
    | _ ->
        Console.WriteLine "Bad input"
        readUserInput msg


let load (mem: int array) operandAddress paramMode =
    match paramMode with
    | ParamMode.Position -> mem.[mem.[operandAddress]]
    | ParamMode.Immediate -> mem.[operandAddress]
    | x -> sprintf "Unexpected ParamMode: %d" (int x) |> ArgumentException |> raise


let store (mem: int array) operandAddress value =
    mem.[mem.[operandAddress]] <- value


let decodeInstruction instruction =
    instruction % 100 |> enum<OpCode>,
    [|
        instruction / 100 % 10 |> enum<ParamMode>
        instruction / 1000 % 10 |> enum<ParamMode>
        instruction / 10000 % 10 |> enum<ParamMode>
    |]


let binaryOp load store op =
    op (load 0) (load 1) |> store 2


let execute (mem: int array) pos =
    let opCode, paramModes = decodeInstruction mem.[pos]
    let load idx = load mem (pos + idx + 1) paramModes.[idx]
    let store idx = store mem (pos + idx + 1)
    let binaryOp = binaryOp load store
    match opCode with
        | OpCode.Add -> binaryOp (+); Increment 4
        | OpCode.Multiply -> binaryOp (*); Increment 4
        | OpCode.Input -> readUserInput "Input: " |> store 0; Increment 2
        | OpCode.Output -> load 0 |> printfn "Output: %d"; Increment 2
        | OpCode.Halt -> Halt mem.[0]
        | x -> Abort (sprintf "Invalid op code %d at %d. Aborting." (int x) pos)


let rec interpret (memory: int array) instructionPtr =
    match execute memory instructionPtr with
    | Increment offset -> interpret memory (instructionPtr + offset)
    | Halt result -> Ok result
    | Abort msg -> Error msg


let rec bruteForceSolve (memory: int array) instructionPtr result noun verb =
    let mem = Array.copy memory
    mem.[1..2] <- [| noun; verb |]
    match interpret mem instructionPtr with
    | Ok res when res = result -> Ok (noun * 100 + verb)
    | _ when noun < 99 || verb < 99 ->
        let nextVerb = (verb + 1) % 100
        let nextNoun = if nextVerb > verb then noun else noun + 1
        bruteForceSolve memory instructionPtr result nextNoun nextVerb
    | _ -> Error "Program could not be solved."


let solve (memory: int array) instructionPtr result =
    bruteForceSolve memory instructionPtr result 0 0

File: rust/int_code/src/interpreter.rs
use num_derive::FromPrimitive;
use num_traits::FromPrimitive;
use std::{fmt::Debug, io::Write};

pub fn interpret(mem: &mut [i32], pos: usize) -> Result<i32, String> {
    let mut prog = decode(mem, pos)?;
    match prog.op_code {
        OpCode::Add => prog.apply_binary(std::ops::Add::add),
        OpCode::Multiply => prog.apply_binary(std::ops::Mul::mul),
        OpCode::Input => {
            let val = read_user_input("Input: ");
            prog.store(0, val);
            prog.ins_ptr += 2;
        }
        OpCode::Output => {
            println!("Output: {}", prog.load(0));
            prog.ins_ptr += 2;
        }
        OpCode::JumpIfTrue => prog.jump(true),
        OpCode::JumpIfFalse => prog.jump(false),
        OpCode::LessThan => prog.apply_binary(|a, b| if a < b { 1 } else { 0 }),
        OpCode::EqualsTo => prog.apply_binary(|a, b| if a == b { 1 } else { 0 }),
        OpCode::Halt => return Ok(mem[0]),
    }

    interpret(prog.memory, prog.ins_ptr)
}

pub fn solve(mem: &mut [i32], pos: usize, result: i32) -> Result<i32, String> {
    for noun in 0..100 {
        for verb in 0..100 {
            let mut tmp_mem = mem.to_vec();
            tmp_mem[1] = noun;
            tmp_mem[2] = verb;
            if interpret(&mut tmp_mem, pos) == Ok(result) {
                return Ok(noun * 100 + verb);
            }
        }
    }
    Err("Program could not be solved.".to_string())
}

#[derive(FromPrimitive, Debug)]
enum OpCode {
    Add = 1,
    Multiply = 2,
    Input = 3,
    Output = 4,
    JumpIfTrue = 5,
    JumpIfFalse = 6,
    LessThan = 7,
    EqualsTo = 8,
    Halt = 99
}

#[derive(FromPrimitive, Copy, Clone, Debug)]
enum ParamMode {
    Positional = 0,
    Immediate = 1
}
struct OpState<'a> {
    op_code: OpCode,
    param_modes: [ParamMode; 3],
    memory: &'a mut [i32],
    ins_ptr: usize,
}

impl<'a> OpState<'a> {
    fn load(&self, index: usize) -> i32 {
        let pos = self.ins_ptr + index + 1;
        match self.param_modes[index] {
            ParamMode::Positional => self.memory[self.memory[pos] as usize],
            ParamMode::Immediate => self.memory[pos]
        }
    }

    fn store(&mut self, index: usize, value: i32) {
        let pos = self.ins_ptr + index + 1;
        match self.param_modes[index] {
            ParamMode::Positional => self.memory[self.memory[pos] as usize] = value,
            ParamMode::Immediate => self.memory[pos] = value
        };
    }

    fn jump(&mut self, test: bool) {
        self.ins_ptr = match self.load(0) != 0 {
            x if x == test => self.load(1) as usize,
            _ => self.ins_ptr + 3
        };
    }

    fn apply_binary<T: Into<i32>>(&mut self, op: fn(i32, i32) -> T) {
        let res = op(self.load(0), self.load(1));
        self.store(2, res.into());
        self.ins_ptr += 4;
    }
}

impl<'a> Debug for OpState<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("instruction: {}, op-code: {:?}, param-modes: {:?}", self.ins_ptr, self.op_code, self.param_modes))
    }
}

fn decode(memory: &mut [i32], ins_ptr: usize) -> Result<OpState, String> {
    let mut instruction = memory[ins_ptr];
    let op_code = OpCode::from_i32(instruction % 100).ok_or(format!("Invalid op-code {} at {}. Aborting.", instruction % 100, ins_ptr))?;

    let mut param_modes: [ParamMode; 3] = [ParamMode::Positional; 3];
    instruction /= 100;
    for (i, mode) in param_modes.iter_mut().enumerate() {
        *mode = ParamMode::from_i32(instruction % 10).ok_or(format!("Invalid parameter mode {} for parameter {} at {}. Aborting.", instruction % 10, i, ins_ptr))?;
        instruction /= 10;
    };

    Ok(OpState{op_code, param_modes, memory, ins_ptr})
}

fn read_user_input(msg: &str) -> i32 {
    print!("{}", msg);
    let _ = std::io::stdout().flush();
    let mut buf = String::new();
    std::io::stdin().read_line(&mut buf).expect("Unable to read from stdin.");
    match buf.trim().parse() {
        Ok(x) => x,
        Err(_) => {
            eprintln!("Invalid input.");
            read_user_input(msg)
        }
    }
}
